test/hotspot/jtreg/containers/docker/TestMemoryAwareness.java

```diff
@@ -96,6 +100,32 @@ private static void testMemoryLimit(String valueToSet, String expectedTraceValue
             .shouldMatch("Memory Limit is:.*" + expectedTraceValue);
     }
 
+    // JDK-8292083
+    // Ensure that Java ignores container memory limit values above the host's physical memory.
+    //
+    // let the host's physical memory be P; request 2P memory in the container.
+    // set java's InitialRAMPercentage to 25%, and check the calculated InitialHeapSize
+    // to see if that was calculated relative to P (P/4) or 2P (P/2).
+    private static void testBadMemoryLimit()
```

Comment:
Could we give this test a better name? How about `testContainerMemExceedsPhysical`? Anyway, we could perhaps write a `WhiteBox` based one which asserts that the `os::physical_memory()` never exceeds the host memory. See `WhiteBox.printOsInfo()` for an example as to how that's done.

As to the code-path in `OSContainer::init` we could assert some log lines with `-Xlog:os+container=trace`. For example for the `Container memory limit exceeded or equal to physical memory!` line to be present. Together with the `os::physical_memory()` whitebox test this should cover all bases. Thoughts?

Reply:
> Could we give this test a better name? How about `testContainerMemExceedsPhysical`?

Of course - that name is good. 

>  Anyway, we could perhaps write a WhiteBox based one which asserts that the os::physical_memory() never exceeds the host memory. See WhiteBox.printOsInfo() for an example as to how that's done.

Do you mean as well as, or instead of, the `OperatingSystemMXBean` approach I've currently taken? (I'm aware it's a bit of a smell. I did test preliminary backports to 11u and 8u, and this is one area that isn't clean)

Looking at `printOsInfo`: is your suggestion to call that and inspect the output, or extend `WhiteBox` itself with a physical memory limit method? I lean towards the latter

Reply:
> > Anyway, we could perhaps write a WhiteBox based one which asserts that the os::physical_memory() never exceeds the host memory. See WhiteBox.printOsInfo() for an example as to how that's done.
> 
> Do you mean as well as, or instead of, the `OperatingSystemMXBean` approach I've currently taken? (I'm aware it's a bit of a smell. I did test preliminary backports to 11u and 8u, and this is one area that isn't clean)

Instead of. OperatingSystemMXBean itself would return the container value but depends on the Java lib code. A cleaner way would be to just use `WhiteBox.physicalMemory()` or some such.
 
> Looking at `printOsInfo`: is your suggestion to call that and inspect the output, or extend `WhiteBox` itself with a physical memory limit method? I lean towards the latter

The latter.

Reply:
With the test as it stands now, there are two Java processes in question: the outer-most test runner (jtreg invoked), which is using OperatingSystemMXBean to establish the host system's real physical RAM. This Java process does not have WhiteBox loaded, I don't think. This process is not subject to a cgroups memory limitation, unless the invoker has set one.

The second java process is run within a docker container by the test, with memory constraints. This launches WhiteBox.

I could extend WhiteBox to report the real physical RAM size (`os::Linux::physical_memory`), but the inner process would simply use that to print, for the outer process to match in the logs. Simpler IMHO would be to add a trace log line reporting the physical RAM size. Neither approach resolves the use of OperatingSystemMXBean in the outer process.

(extending WhiteBox to report (`os::Linux::physical_memory`) also means importing "os_linux.hpp" in "whitebox.cpp", and moving that method from `protected` to `public`. The method we write also needs to be substantially IFDEF'd to Linux-only, and return an error value for everyone else.)

Doing that means we are not comparing the view of system RAM between the inner and outer processes, and instead relying entirely on what the inner process reports. Is that a problem?

Reply:
Pre-empting your reply (and if I've guessed wrong I'll push a revert): I've re-written the test to be simpler and not use OperatingSystemMXBean. It invokes two separate docker/java instances: the first to capture physical memory size; the second to set a bogus memory limit and ensure that it's caught. (ff57cf418e39fdb619bd6fe8f8d6bcfca953380a)